File: TypeScript/permutations.ts
function permuteArray(permutations: number[][], aux: number[], nums: number[]) {
    if(aux.length == nums.length) {
        permutations.push([...aux]);
        return;
    } else {
        for(const num of nums) {
            if(aux.includes(num)) {
                continue;
            }

            aux.push(num);
            permuteArray(permutations, aux, nums);
            aux.pop();
        }
    }
}
function permute(nums: number[]): number[][] {
    let permutations = []
    permuteArray(permutations, [], nums);
    return permutations;
};

File: Python/permutations.py
class Solution:
    def permuteArray(self, permutations: List[List[int]], aux: List[int], nums: List[int]):
        if(len(aux) == len(nums)):
            permutations.append(aux[:])
            return
        else:
            for num in nums:
                if(num in aux):
                    continue
                aux.append(num)
            
                self.permuteArray(permutations, aux, nums)
            
                aux.pop()

    def permute(self, nums: List[int]) -> List[List[int]]:
        permutations = []
        self.permuteArray(permutations, [], nums)

        return permutations